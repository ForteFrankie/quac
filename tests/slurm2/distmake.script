#!/bin/bash

# Test that distributed make over ssh works in a SLURM allocation.

# Copyright (c) 2012-2013 Los Alamos National Security, LLC, and others.

. ./environment.sh

cd $DATADIR
exec 2>&1

## SSH multiplexing test ##
x sshrot --muxstart
# Verify one master to each node.
sshinfo $SLURM_NODE_A
sshinfo $SLURM_NODE_B
# Run three tasks.
y "sshrot 'echo a && sleep 1 && echo b' &"
y "sshrot 'echo a && sleep 1 && echo b' &"
y "sshrot 'echo a && sleep 1 && echo b' &"
# Wait for output.
x sleep 0.75
# Verify two connections to node A, one to node B.
sshinfo $SLURM_NODE_A
sshinfo $SLURM_NODE_B
# Wait for output.
x sleep 1.25
# Stop multiplexing.
x sshrot --muxstop
# Verify that everything is gone.
sshinfo $SLURM_NODE_A
sshinfo $SLURM_NODE_B

## SSH scale test ##
x sshrot --muxstart
for i in {1..18}; do
    # Note: I don't actually want all 18 sshrot invocations in the output log,
    # but if I remove the y wrapper, then the ssh invocations show up in ps
    # output as sshrot instead of ssh, breaking sshinfo function. ???
    y "sshrot 'sleep 1.5 && echo c' &"
done
x sleep 0.5
sshinfo $SLURM_NODE_A
sshinfo $SLURM_NODE_B
x sleep 2.0
x sshrot --muxstop

# QUACreduce to multiple nodes test
x sshrot --muxstart
echo -e 'foo bar baz\nfoo foo' > foo1.txt
echo -e 'bar' > foo2.txt
quacreduce --dist \
           --map 'tr "[:blank:]" "\n"' \
           --reduce 'uniq -c > out/%(RID)' \
           --partitions 3 \
           foo*.txt
x make -j2 --quiet
y 'cat out/* | sort'
x sshrot --muxstop
