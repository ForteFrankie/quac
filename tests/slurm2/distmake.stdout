$ sshrot --muxstart
ssh clients:      0
ssh masters:      1
control sockets:  1
TCP connections:  1
ssh clients:      0
ssh masters:      1
control sockets:  1
TCP connections:  1
$ (sshrot 'echo a && sleep 1 && echo b' &)
$ (sshrot 'echo a && sleep 1 && echo b' &)
$ (sshrot 'echo a && sleep 1 && echo b' &)
$ sleep 0.75
a
a
a
ssh clients:      2
ssh masters:      1
control sockets:  1
TCP connections:  1
ssh clients:      1
ssh masters:      1
control sockets:  1
TCP connections:  1
$ sleep 1.25
b
b
b
$ sshrot --muxstop
Exit request sent.
Exit request sent.
ssh clients:      0
ssh masters:      0
control sockets:  0
TCP connections:  0
ssh clients:      0
ssh masters:      0
control sockets:  0
TCP connections:  0
$ sshrot --muxstart
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ (sshrot 'sleep 1.5 && echo c' &)
$ sleep 0.5
ssh clients:      9
ssh masters:      1
control sockets:  1
TCP connections:  1
ssh clients:      9
ssh masters:      1
control sockets:  1
TCP connections:  1
$ sleep 2.0
c
c
c
c
c
c
c
c
c
c
c
c
c
c
c
c
c
c
$ sshrot --muxstop
Exit request sent.
Exit request sent.
$ sshrot --muxstart
$ make -j2 --quiet
$ (cat out/* | sort)
      1 baz
      2 bar
      3 foo
$ sshrot --muxstop
Exit request sent.
Exit request sent.
